auth_mellon_util.c

```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
+{
+    switch(token_type) {
+    case TOKEN_WHITESPACE:       return "WHITESPACE";
+    case TOKEN_SEMICOLON:        return "SEMICOLON";
+    case TOKEN_COMMA:            return "COMMA";
+    case TOKEN_EQUAL:            return "EQUAL";
+    case TOKEN_IDENTIFIER:       return "IDENTIFIER";
+    case TOKEN_DBL_QUOTE_STRING: return "DBL_QUOTE_STRING";
+    default:                     return "unknown";
+    }
+}
+
+static void dump_tokens(request_rec *r, apr_array_header_t *tokens)
+{
+    apr_size_t i;
+    
+    for (i = 0; i < tokens->nelts; i++) {
+        Token token = APR_ARRAY_IDX(tokens, i, Token);
+        ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r,
+                      "token[%2zd] %s \"%s\" offset=%lu len=%lu ", i,
+                      token_type_str(token.type), token.str,
+                      token.offset, token.len);
+    }
+}
+#endif
+
+/* Break a string into a series of tokens
  *
- * A PAOS header must be composed of 2 values seperated by a semicolon.
+ * Given a string return an array of tokens. If the string cannot be
+ * successfully parsed an error string is returned at the location
+ * specified by the error parameter, if error is NULL then the parsing
+ * was successful. If an error occured the returned array of tokens
+ * will include all tokens parsed up until where the unrecognized
+ * input occurred. The input str is never modified.
  *
- * The first value must be a version declaration in the form ver="xxx"
- * (note the version string must be in double quotes).
+ * Parameters:
+ *  pool              memory allocation pool
+ *  str               input string to be parsed.
+ *  ignore_whitespace if True whitespace tokens are not returned
+ *  error             location where error string is returned
+ *                    if NULL no error occurred
+ * Returns:           array of Token objects
+ */
+
+#define TOKEN_INIT(t) memset(&t, 0, sizeof(t))
+
+#define PUSH_TOKEN(token_type, start, end)              \
+{                                                       \
+    const char *s = (start);                            \
+    const char *e = (end);                              \
+    apr_size_t offset = s - str;                        \
+                                                        \
+    if (token_type == TOKEN_DBL_QUOTE_STRING) {         \
+        /* do not include quotes in token value */      \
+        s++; e--;                                       \
+    }                                                   \
+                                                        \
+    token.type = token_type;                            \
+    token.len = e-s;                                    \
+    token.offset = offset;                              \
+    token.str = apr_pstrmemdup(pool, s, token.len);     \
+    APR_ARRAY_PUSH(tokens, Token) = token;              \
+                                                        \
+    TOKEN_INIT(token);                                  \
```

Comment:
Instead of using a single token variable across all instances of this `#define`, couldn't you simply change the code to something like:

```
Token *token = apr_array_push(tokens);
token->type = token_type;
[...]
```